### ISSUE TITLE
Rotation invariance

### ISSUE THREAD

**Issue by cv2000** [Celebrandil/CudaSift#45]
Hi,
After replacing OpenCV's implementation of SIFT with CudaSift in my code I noticed that it could no longer handle much rotation between the images being matched. Is this a known limitation to CudaSift, or am I missing something?

When testing on mainSift.cpp and the sample data I get the following on the original images img1.png and img2.png:
Number of original features: 1818 1978
Number of matching features: 792 801 43.5644% 1 3

When rotating one of the images 180 degrees I get the following:
Number of original features: 1818 2040
Number of matching features: 261 400 14.3564% 1 3

**Comment by Celebrandil** [comment]
No, there should be no such limitation, since all operations should be rotationally invariant (at least if we disregard the discretization of pixels). For a 180 degree rotation, it should be exactly the same. I can confirm your observation though. Thank you! It's often hard to confirm the existence of a bug when you are working with noisy data. I'll try to find the problem as soon as I can.

**Comment by Celebrandil** [comment]
I've found an explanation for your observations. There was an incorrect shift of (0.5, 0.5) pixels when reading image data to create the descriptors. If you match two different images that are similarly oriented, it doesn't matter that much, but when one image is rotated 180 degrees it does. I'll provide a fix when I've excluded other possible reasons.

Number of original features: 1911 1931
Number of matching features: 1414 1440 73.9927% 1 3

**Comment by Karol-G** [comment]
Are you still working on the fix? Or did it turn out to be more complicated?

**Comment by Celebrandil** [comment]
I have a fix, but I have to be super careful not to introduce any new problems.

**Comment by Celebrandil** [comment]
Could you test the new version in the new Pascal branch?

**Comment by cv2000** [comment]
Thank you! I did some quick tests and now I get similar results when rotating one of the images.

Original images:
Number of original features: 1911 2086
Number of matching features: 839 856 43.9037% 1 3

When rotating one of the images 180 degrees
Number of original features: 1911 2086
Number of matching features: 784 909 41.0256% 1 3

In my code I also notice a big improvement when it comes to rotation. Still not comparable to the accuracy and robustness I get from OpenCV's implementation in general, but I have not played around with the parameters too much. Do you know if this is expected or do I just have to find the right parameters?

**Comment by Celebrandil** [comment]
No, I haven't done any systematic comparison to the OpenCV implementation, but I would love to see that done some day. Originally, I compared my earlier versions to Lowe's own implementation and for each new version I've tried to ensure that you don't lose accuracy along the way. Worth noting is that the more feature you extract, the smaller fraction of features can be matched. Finding the sweet spot typically depends on your application.
